File: core/src/main/java/io/tn/core/lang/DTimer.java
package io.tn.core.lang;

import java.time.*;
import java.time.temporal.ChronoUnit;
import java.util.Date;

/**
 * 时间工具类
 *
 * @author TrueNine
 * @since 2022-12-16
 */
public class DTimer {
  public static final String MILLIS = "SSS";
  public static final String DATE = "yyyy-MM-dd";
  public static final String TIME = "HH:mm:ss";
  public static final String DATETIME = DATE + " " + TIME;
  public static final String TIME_M = TIME + " " + MILLIS;
  public static final String DATETIME_M = DATETIME + " " + MILLIS;

  public static final String DATE_S = "yyyy/MM/dd";
  public static final String DATETIME_S = DATE_S + " " + TIME;
  public static final String DATETIME_M_S = DATETIME_S + " " + MILLIS;


  public static void sleepMillis(long millis) {
    try {
      Thread.sleep(millis);
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
  }


  public static void sleep(long second) {
    sleepMillis(second * 1000);
  }

  public static void sleepOne() {
    sleep(1);
  }

  public static Duration ofSecondDuration(long second) {
    return Duration.of(second, ChronoUnit.SECONDS);
  }

  public static Duration ofMillisDuration(long millis) {
    return Duration.of(millis, ChronoUnit.MILLIS);
  }

  public static Date plusDate(long plusMillis) {
    return plusMillis(System.currentTimeMillis(), plusMillis);
  }

  public static Date plusMillis(long current, long plusMillis) {
    return new Date(current + plusMillis);
  }

  public static Date localTimeToDate(LocalTime lt) {
    var meta = LocalDate.of(1970, 1, 1);
    var ldt = lt.atDate(meta);
    return Date.from(
        ldt.atZone(ZoneId.of("GMT"))
            .toInstant()
    );
  }

  public static Date localDateToDate(LocalDate ld) {
    return Date.from(
        ld.atStartOfDay().atZone(ZoneId.systemDefault())
            .toInstant()
    );
  }

  public static Date localDatetimeToDate(LocalDateTime ldt) {
    return Date.from(
        ldt.atZone(ZoneId.systemDefault())
            .toInstant()
    );
  }
}

File: core/src/main/kotlin/net/yan100/compose/core/lang/DTimer.kt
package net.yan100.compose.core.lang

import net.yan100.compose.core.lang.DTimer.ZONE_GMT
import java.time.*
import java.time.temporal.ChronoUnit
import java.util.*

/**
 * 时间工具类
 *
 * @author TrueNine
 * @since 2022-12-16
 */
object DTimer {
    const val ZONE_UTC = "UTC"
    const val ZONE_GMT = "Etc/GMT"
    const val MILLIS = "SSS"
    const val DATE = "yyyy-MM-dd"
    const val TIME = "HH:mm:ss"
    const val DATETIME = DATE + " " + TIME
    const val DATETIME_M = DATETIME + " " + MILLIS
    const val TIME_M = TIME + " " + MILLIS
    const val DATE_S = "yyyy/MM/dd"
    const val DATETIME_S = DATE_S + " " + TIME
    const val DATETIME_M_S = DATETIME_S + " " + MILLIS

    @JvmStatic
    fun sleepMillis(millis: Long) {
        try {
            Thread.sleep(millis)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }

    @JvmStatic
    fun sleep(second: Long) {
        sleepMillis(second * 1000)
    }

    @JvmStatic
    fun getDurationBySecond(second: Long): Duration {
        return Duration.of(second, ChronoUnit.SECONDS)
    }

    @JvmStatic
    fun getDurationByMillis(millis: Long): Duration {
        return Duration.of(millis, ChronoUnit.MILLIS)
    }

    @JvmStatic
    fun plusMillisFromCurrent(plusMillis: Long): Date {
        return plusMillis(System.currentTimeMillis(), plusMillis)
    }

    @JvmStatic
    fun plusMillis(current: Long, plusMillis: Long = System.currentTimeMillis()): Date {
        return Date(current + plusMillis)
    }

    @JvmStatic
    fun localTimeToDate(lt: LocalTime, zoneId: ZoneId = ZoneId.of("GMT")): Date {
        val meta = LocalDate.of(1970, 1, 1)
        val ldt = lt.atDate(meta)
        return Date.from(
            ldt.atZone(zoneId)
                .toInstant()
        )
    }

    @JvmStatic
    fun localDateToDate(ld: LocalDate, zoneId: ZoneId = ZoneId.systemDefault()): Date {
        return Date.from(
            ld.atStartOfDay().atZone(zoneId)
                .toInstant()
        )
    }

    @JvmStatic
    fun localDatetimeToDate(ldt: LocalDateTime, zoneId: ZoneId = ZoneId.systemDefault()): Date {
        return Date.from(
            ldt.atZone(zoneId)
                .toInstant()
        )
    }

    @JvmStatic
    fun dateToLocalDatetime(date: Date, zoneId: ZoneId = ZoneId.systemDefault()): LocalDateTime {
        return date.toInstant()
            .atZone(zoneId)
            .toLocalDateTime()
    }

    @JvmStatic
    fun millisToLocalDateTime(millis: Long, zoneId: ZoneId = ZoneId.systemDefault()): LocalDateTime {
        return Instant.ofEpochMilli(millis).atZone(zoneId).toLocalDateTime()
    }

    @JvmStatic
    fun dateToLocalDate(date: Date, zoneId: ZoneId = ZoneId.systemDefault()): LocalDate {
        return date.toInstant().atZone(zoneId)
            .toLocalDate()
    }

    @JvmStatic
    fun millisToLocalDate(millis: Long, zoneId: ZoneId = ZoneId.systemDefault()): LocalDate {
        return Instant.ofEpochMilli(millis).atZone(zoneId).toLocalDate()
    }

    @JvmStatic
    fun dateToLocalTime(date: Date, zoneId: ZoneId = ZoneId.of(ZONE_GMT)): LocalTime {
        return date.toInstant().atZone(zoneId).toLocalTime()
    }

    @JvmStatic
    fun millisToLocalTime(millis: Long, zoneId: ZoneId = ZoneId.of(ZONE_GMT)): LocalTime {
        return Instant.ofEpochMilli(millis).atZone(zoneId).toLocalTime()
    }
}

fun Long.toDate(): Date {
    return Date(this)
}

fun Long.toLocalDateTime(zoneId: ZoneId = ZoneId.systemDefault()): LocalDateTime {
    return DTimer.millisToLocalDateTime(this, zoneId)
}

fun Long.toLocalDate(zoneId: ZoneId = ZoneId.systemDefault()): LocalDate {
    return DTimer.millisToLocalDate(this)
}

fun Long.toLocalTime(zoneId: ZoneId = ZoneId.of(ZONE_GMT)): LocalTime {
    return DTimer.millisToLocalTime(this)
}


fun Date.toLong(): Long {
    return this.time
}

fun Date.toLocalDatetime(zoneId: ZoneId = ZoneId.systemDefault()): LocalDateTime {
    return DTimer.dateToLocalDatetime(this, zoneId)
}

fun Date.toLocalDate(zoneId: ZoneId = ZoneId.systemDefault()): LocalDate {
    return DTimer.dateToLocalDate(this, zoneId)
}

fun Date.toLocalTime(zoneId: ZoneId = ZoneId.of(ZONE_GMT)): LocalTime {
    return DTimer.dateToLocalTime(this, zoneId)
}

fun LocalTime.toDate(zoneId: ZoneId = ZoneId.of(ZONE_GMT)): Date {
    return DTimer.localTimeToDate(this, zoneId)
}

fun LocalDate.toDate(zoneId: ZoneId = ZoneId.systemDefault()): Date {
    return DTimer.localDateToDate(this, zoneId)
}

fun LocalDateTime.toDate(zoneId: ZoneId = ZoneId.systemDefault()): Date {
    return DTimer.localDatetimeToDate(this, zoneId)
}
